File: api/internal/app/repository/otomo.go
package repository

import (
	"context"
	"otomo/internal/app/interfaces/repo"
	"otomo/internal/app/model"
	"otomo/internal/pkg/errs"

	"cloud.google.com/go/firestore"
)

var _ repo.OtomoRepository = (*OtomoRepository)(nil)

type OtomoRepository struct {
	fsClient *firestore.Client
}

func NewOtomoRepository(fsClient *firestore.Client) *OtomoRepository {
	return &OtomoRepository{
		fsClient: fsClient,
	}
}

func (or *OtomoRepository) Save(
	ctx context.Context,
	otomo *model.Otomo,
) error {
	doc := or.getDoc(ctx, otomo.UserID)
	_, err := doc.Set(ctx, otomo)
	return err
}

func (or *OtomoRepository) GetByID(
	ctx context.Context,
	id model.UserID,
) (*model.Otomo, error) {
	doc := or.getDoc(ctx, id)
	snapshot, err := doc.Get(ctx)
	if err != nil {
		return nil, ifCodesNotFoundReturnErrsNotFound(
			err, errs.DomainOtomo, errs.FieldUserID)
	}

	var otomo = &model.Otomo{}
	if err := snapshot.DataTo(otomo); err != nil {
		return nil, err
	}
	return otomo, nil
}

func (r *OtomoRepository) getDoc(
	ctx context.Context,
	userID model.UserID,
) *firestore.DocumentRef {
	return r.fsClient.Doc(GetOtomoDocPath(userID))
}

File: frontend/lib/controllers/otomo.dart
import 'package:injectable/injectable.dart';
import 'package:otomo/domains/entities/otomo.dart';
import 'package:otomo/domains/repo/otomo.dart';
import 'package:otomo/tools/logger.dart';

@injectable
class OtomoControllerImpl {
  const OtomoControllerImpl(this._repository);

  final OtomoRepository _repository;

  Future<void> save(Otomo otomo) {
    logger.info('OtomoControllerImpl.save');
    return _repository.save(otomo);
  }

  Future<Otomo> get(String userId) {
    logger.info('OtomoControllerImpl.get');
    return _repository.get(userId);
  }
}
